src/librbd/ImageCtx.cc

```diff
@@ -813,6 +813,16 @@ class SafeTimerSingleton : public SafeTimer {
       RBD_QOS_WRITE_BPS_THROTTLE,
       config.get_val<uint64_t>("rbd_qos_write_bps_limit"),
       config.get_val<uint64_t>("rbd_qos_write_bps_burst"));
+
+    // the write buffer might not be stable so we cannot zero-copy if the
+    // data frame is being checked for integrity
+    if (!disable_zero_copy &&
+        (config.get_val<bool>("ms_crc_data") ||
+         config.get_val<bool>("ms_msgr2_sign_messages") ||
+         config.get_val<bool>("ms_msgr2_encrypt_messages"))) {
```

Comment:
ms_msgr2_{sign,encrypt}_messages are outdated, I've submitted https://github.com/ceph/ceph/pull/31818 to remove them.  The option that is actually used is ms_client_mode, a list of modes in order of preference.

I'm not sure we should parse it in librbd though, perhaps there is a way we can query the negotiated mode, something like AuthConnectionMeta::con_mode?  @liewegas?

Reply:
> I'm not sure we should parse it in librbd though, perhaps there is a way we can query the negotiated mode, something like AuthConnectionMeta::con_mode? @liewegas?

That wouldn't work since nothing would stop one OSD connection from being encrypted and another not if mid-upgrade. Therefore, it would need to just disable zero-copy if enabled in the `ms_client_mode`. i.e. the only way to get zero-copy writes would be in benchmark tests when all these things are turned off to get snazzy numbers.

Reply:
Yeah, not sure what I was thinking.  Checking pretty much anything on a connection is racy, I should know as I've run into it more than once in the kernel...

Reply:
I could tweak the buffer::list/ptr/raw to add the concept along the lines of `make_stable` (like the old `make_shareable`) which could be invoked down in the msgr layer before writing the payload if CRCs/encryption/etc are enabled. librbd could then flag zero-copy buffer::raw as unstable and delay the copy until the last possible moment.

Reply:
... after a quick scan of the V2 protocol code, it seems like you are either in secure more or CRC mode w/ no option to disable both, so basically we would always need to disable zero-copy writes.

Reply:
I just added a new config option to let clients opt-in if they know they will not issue unstable writes.